cranelift-filetests/src/function_runner.rs

```diff
@@ -1,53 +1,275 @@
+use core::fmt::{self, Display, Formatter};
 use core::mem;
+use core::str::FromStr;
 use cranelift_codegen::binemit::{NullRelocSink, NullStackmapSink, NullTrapSink};
-use cranelift_codegen::ir::Function;
+use cranelift_codegen::ir::{types, AbiParam, ArgumentPurpose, ConstantData, Function};
 use cranelift_codegen::isa::TargetIsa;
 use cranelift_codegen::{settings, Context};
 use cranelift_native::builder as host_isa_builder;
-use memmap::MmapMut;
+use cranelift_reader::Parser;
+use memmap::{Mmap, MmapMut};
 
 /// Run a function on a host
 pub struct FunctionRunner {
     function: Function,
     isa: Box<dyn TargetIsa>,
+    action: FunctionRunnerAction,
+}
+
+/// Build a function runner
+pub struct FunctionRunnerBuilder {
+    function: Function,
+    isa: Option<Box<dyn TargetIsa>>,
+    action: FunctionRunnerAction,
+}
+
+#[derive(Clone, Debug)]
+pub enum FunctionRunnerAction {
+    Print,
+    Test,
+    TestWithResult { operator: Operator, expected: Value },
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum Operator {
+    EQ,
+    NEQ,
+}
+
+impl Display for Operator {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        f.write_str(match self {
+            Self::EQ => "==",
+            Self::NEQ => "!=",
+        })
+    }
+}
+
+impl FromStr for Operator {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "==" => Ok(Self::EQ),
+            "!=" => Ok(Self::NEQ),
+            _ => Err(format!("Unsupported operator {}", s)),
+        }
+    }
+}
+
+macro_rules! try_into_int {
+    ($type: ty, $name: ident) => {
+        struct $name {}
+        impl $name {
+            fn try_into(d: ConstantData) -> Result<$type, String> {
```

Comment:
Instead of making an ad-hoc `try_into` method, we should follow convention and implement the `TryFrom` trait, which gives us a `try_into` via the blanket impl `TryInto<T> for U` where `U: TryFrom<T>`. For details, see: https://doc.rust-lang.org/stable/std/convert/trait.TryInto.html and https://doc.rust-lang.org/stable/std/convert/trait.TryFrom.html

If I'm reading this code right, it would be something like this:

```rust
struct $name(ConstantData);

impl TryFrom<$name> for $type {
    // ...
}
```

But this also makes me wonder why we don't implement `TryFrom<ConstantData> for $type` directly instead of creating all these temporary `ConstantData$Type` types. Is that something that you explored at all? I think it would be the best course to take, unless it doesn't work for some reason or I am overlooking something else.